course-design.md

```diff
@@ -0,0 +1,214 @@
+# Course design components for *RSE in R*
+
+## Basic design components
+
+These are to develop a compass for developing and structuring the material.
+Many may have a strong focus on "in class" etc, 
+but this is to give a direction on how to write those sections
+(e.g. having the perspective that an instructor will be reading these sections).
+
+- **Learning outcome/goal**: Ultimately, to be able to create an R package, that:
+    - Passes checks and requirements to (potentially) be submitted to CRAN
+    - Incorporates best practices in package development 
+    (unit testing, continuous integration, version control)
+    - Has code that is written and developed in a readable, clean, and
+    style-guide consistent format, so others can more easily read it
+    - Is designed to involve the community, such as gaining potential contributors
+    (including licensing, code of conduct, contributing guidelines, 
+    well described README)
+    - Has user-friendly (i.e. targeted to a novice to intermediate user)
+    documentation and tutorials on how to use the package (e.g. a nice website)
+    - Is adequately structured to make package management and maintenance easier
+    (e.g. with Issue Templates, Project Boards, Labels, contributing guidelines)
+
+- **Teaching/learning activities**:
+    - In class portion: 
+        - Some lectures for those material that aren't coding heavily
+        (e.g. for lessons to involve community, being a good community member)
+        - Live coding (with instructor)
+        - Exercises that expand on material covered with instructor that 
+        solve a provided problem, with or without scaffolding
+        - Could include dedicated group work sessions after the teamwork lessons
+        (at end), with instructor and helpers available in class
+    - Out of class: 
+        - Group work/project creating an R package based on a set of (given?) problems
+            - Or include several potential projects that are either individual
+            or group based so that instructor or learner has options to choose from
+            based on the situation, class structure, and other needs
+            - Maybe provide several potential problem sets as starting points? Or 
+            get the learners to determine their own? (this might be difficult) 
+            Optional is to choose a few simple existing packages (that don't
+            follow these guidelines from the book) and get them to re-make it
+            using the principles (and maybe later ask package maintainer if they
+            would like the features as a PR?)
```

Comment:
This seems difficult since we would need to find quite many and also keep this list updated. Or do you mean that each instructor would find this as they teach the class?

Reply:
I'm thinking more as a starting point. Instructors can decide the specifics of the assignment on their own, but we will at least give them something to start thinking about. @cwickham's comment at the bottom sums it up.

Reply:
Choosing existing packages that could be improved would be a cool community service. I agree with Joel that this would work best with frequent updating, but even if the list was not updated, it would be neat for future instructors to see some before-and-after examples from previous classes.